### PR TITLE
[dhctl] [cloud-provider-openstack] Add Openstack API availability check inside the cluster during deployment

### PR DESCRIPTION
**Pull request by borg-z**
## Description
Added a new `PostCloud` step that is executed only when bootstrapping occurs in a cloud environment. The `PostCloud` step runs after the first master server is created and SSH connectivity is verified. Within `PostCloud`, a check for "access to cloud API from the master host" has been implemented specifically for the cloud providers vSphere and OpenStack.

Two methods are used for this verification:
- **With HTTP Proxy**: An SSH tunnel is established to the master server, forwarding a port to the proxy. A GET request is then made to the cloud API using `localhost:22323` as the proxy.
- **Without HTTP Proxy**: A port is forwarded directly to the cloud API (e.g., `-L 22323:api.selvpc.ru:443`). A GET request is made to `localhost:22323` while specifying the original TLS SNI.

These changes ensure that the master host can successfully communicate with the cloud API.

## Why do we need it, and what problem does it solve?

> At the moment, there is a situation when, when deploying via an external installation container, the first master node is successfully deployed using Terraform, but further VMs and objects are not created from within the cluster because the Openstack API is not available from within the cluster.

Now we check this situation before the cluster is installed

Fixes issue https://github.com/deckhouse/deckhouse/issues/9970

## Why do we need it in the patch release (if we do)?

*(Remove this section if the PR is not intended for a patch release.)*

## What is the expected result?

After applying these changes:
- When bootstrapping in a cloud environment (vSphere or OpenStack), the `PostCloud` step will execute.
- The system will verify access to the cloud API from the master host.
- Successful verification ensures that the master host can communicate with the cloud API.
- In case of verification failure, the bootstrap process will provide clear error messages indicating the inability to access the cloud API.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
```changes
section: dhctl
type: feature
summary: Add PostCloud step to verify cloud API access from master host during bootstrap
impact:
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by borg-z** [comment]
/e2e/abort openstack;Standard;containerd;1.27 11557466600 11557466600-1-con-1-27 /sys/deckhouse-oss/install:pr10311 redos@87.228.36.210

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by borg-z** [comment]
/e2e/abort vsphere;Standard;containerd;1.27 11558671370 11558671370-1-con-1-27 /sys/deckhouse-oss/install:pr10311 ubuntu@10.220.203.26

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by borg-z** [comment]
/e2e/abort vsphere;Standard;containerd;1.27 11569338792 11569338792-1-con-1-27 /sys/deckhouse-oss/install:pr10311 ubuntu@10.220.203.26

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by borg-z** [comment]
/e2e/abort vsphere;Standard;containerd;1.27 11570254051 11570254051-1-con-1-27 /sys/deckhouse-oss/install:pr10311 ubuntu@10.220.203.26

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by borg-z** [comment]
/e2e/abort vsphere;Standard;containerd;1.27 11570747851 11570747851-1-con-1-27 /sys/deckhouse-oss/install:pr10311 ubuntu@10.220.203.87

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by borg-z** [comment]
/e2e/abort static;Static;containerd;1.27 11573793556 11573793556-1-con-1-27 /sys/deckhouse-oss/install:pr10311 astra@192.168.199.155

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by borg-z** [comment]
/e2e/abort static;Static;containerd;1.27 11574224065 11574224065-1-con-1-27 /sys/deckhouse-oss/install:pr10311 astra@192.168.199.175

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/static'...

**Comment by borg-z** [comment]
/e2e/abort static;Static;containerd;1.27 11576007150 11576007150-1-con-1-27 /sys/deckhouse-oss/install:pr10311 astra@192.168.199.105

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by borg-z** [comment]
/e2e/abort static;Static;containerd;1.27 11576007150 11576007150-1-con-1-27 /sys/deckhouse-oss/install:pr10311 astra@192.168.199.105

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/gcp'...

File: dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go [hunk 1]
```diff
@@ -17,10 +17,11 @@ package bootstrap
 import (
 	"encoding/json"
 	"fmt"
-	"github.com/deckhouse/deckhouse/dhctl/pkg/kubernetes"
 	"reflect"
 	"time"
 
+	"github.com/deckhouse/deckhouse/dhctl/pkg/kubernetes"
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go:23]: imports not sorted

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+type CloudApiConfig struct {
+	URL      *url.URL
+	Insecure bool
+	CACert   string
+}
+
+func (pc *Checker) CheckCloudAPIAccessibility() error {
+
+	if app.PreflightSkipCloudAPIAccessibility {
+		log.InfoLn("Checking  Cloud Api is accessible from first master host was skipped (via skip flag)")
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:69]:
```suggestion
		log.InfoLn("Checking Cloud API is accessible from first master host was skipped (via skip flag)")
```

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+		return nil
+	}
+
+	log.DebugLn("Checking if Cloud Api is accessible from first master host")
+	wrapper, ok := pc.nodeInterface.(*ssh.NodeInterfaceWrapper)
+	var tun *frontend.Tunnel
+
+	if !ok {
+		log.InfoLn("Checking if Cloud Api is accessible through proxy was skipped (local run)")
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:78]:
```suggestion
		log.InfoLn("Checking if Cloud API is accessible through proxy was skipped (local run)")
```

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	}
+
+	proxyUrl, noProxyAddresses, err := getProxyFromMetaConfig(pc.metaConfig)
+	if err != nil {
+		return fmt.Errorf("get proxy config: %w", err)
+	}
+
+	cloudApiConfig, err := getCloudApiConfigFromMetaConfig(pc.metaConfig)
+	if err != nil {
+		log.ErrorF("Cannot parse CloudApiConfiguration: %v", err)
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:89]:
```suggestion
		log.ErrorF("Cannot parse Cloud API Configuration: %v", err)
```

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+		log.InfoLn("Checking if Cloud Api is accessible through proxy was skipped (local run)")
+		return nil
+	}
+
+	proxyUrl, noProxyAddresses, err := getProxyFromMetaConfig(pc.metaConfig)
+	if err != nil {
+		return fmt.Errorf("get proxy config: %w", err)
+	}
+
+	cloudApiConfig, err := getCloudApiConfigFromMetaConfig(pc.metaConfig)
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:87]:
```suggestion
	cloudAPIConfig, err := getCloudApiConfigFromMetaConfig(pc.metaConfig)
```

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	}
+	body := string(bodyBytes)
+	statusCode := resp.StatusCode
+	log.DebugF("status, response: %d %s", statusCode, body)
+	// Debug
+
+	return resp, nil
+}
+
+func getCloudApiConfigFromMetaConfig(metaConfig *config.MetaConfig) (*CloudApiConfig, error) {
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:203]:
```suggestion
func getCloudAPIConfigFromMetaConfig(metaConfig *config.MetaConfig) (*CloudApiConfig, error) {
```

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+		return fmt.Errorf("get proxy config: %w", err)
+	}
+
+	cloudApiConfig, err := getCloudApiConfigFromMetaConfig(pc.metaConfig)
+	if err != nil {
+		log.ErrorF("Cannot parse CloudApiConfiguration: %v", err)
+		return err
+	}
+
+	if cloudApiConfig == nil {
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:93]:
```suggestion
	if cloudAPIConfig == nil {
```

File: dhctl/pkg/preflight/check_cloud_api.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to build HTTP client: %w", err)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("HTTP request failed: %w", err)
+	}
+
+	// Debug
```
Comment (name212) [inline dhctl/pkg/preflight/check_cloud_api.go:188]: Strange comment

File: dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go [hunk 9]
```diff
@@ -17,10 +17,11 @@ package bootstrap
...
+
```
Comment (name212) [inline dhctl/pkg/operations/bootstrap/cluster-bootstrapper.go:24]:
```suggestion
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by borg-z** [comment]
/e2e/abort openstack;Standard;containerd;1.27 11705513531 11705513531-1-con-1-27 /sys/deckhouse-oss/install:pr10311 redos@37.9.4.152

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by borg-z** [comment]
Added check output:
```
│ ┌ Checking access to cloud api from master host
│ │ GET https://api.selvpc.ru/identity/v3: 200 OK
│ │ 🎉 Succeeded!
│ └ Checking access to cloud api from master host (0.45 seconds)
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/gcp'...

**Review by name212 — APPROVED** [review]

**Review by sprait — APPROVED** [review]

**Review by aleksey-su — APPROVED** [review]

**Pull request merged**